### PR TITLE
[NPU]:Added support for the layer_norm operator in npu

### PR DESCRIPTION
**Pull request by TianHao324**
## Summary

Added NPU-friendly operator implementations. Previously, only some unsupported writing methods were addressed based on the GPU implementation. However, when the n_col is large, the problem of UB overflow still occurs. This modification has solved this issue, and the performance has improved by 2-3 times compared to the original implementation. However, due to device limitations, the performance is still somewhat lower than that of Hugging Face. We will continue to follow up on the performance issues in the future.

## Testing Done

image

- Hardware Type: Atlas 800I A2
- [x] run `make test` to ensure correctness
- [x] run `make checkstyle` to ensure code style
- [ ] run `make test-convergence` to ensure convergence

### CODE REVIEW

**Comment by TianHao324** [comment]
@Tcc0403 would you mind having a preview?

File: src/liger_kernel/ops/backends/_ascend/ops/layer_norm.py [hunk 1]
```diff
@@ -0,0 +1,681 @@
...
+
+        for i in range(num_groups):
+            start_idx = i * group_size
+            end_idx = min((i + 1) * group_size, num_programs)
+            if start_idx < end_idx:
+                group_sum = tensor[start_idx:end_idx].sum(dim=dim, dtype=torch.float32)
+                groups.append(group_sum)
+
+        if len(groups) > 0:
+            return torch.stack(groups).sum(dim=0, dtype=torch.float32)
```
Comment (Tcc0403) [inline src/liger_kernel/ops/backends/_ascend/ops/layer_norm.py]: Is it hard to pass correctness tests without this helper function?

File: src/liger_kernel/ops/backends/_ascend/ops/layer_norm.py [hunk 2]
```diff
@@ -0,0 +1,681 @@
...
+            wdy = W_block * DY_block
+
+            DX_block = (wdy - (x_hat * c1 + c2)) * rstd
+            tl.store(DX_row_ptr + col_offsets, DX_block.to(X_dtype), mask=mask)
+
+            dW_block = DY_block * x_hat
+            dB_block = DY_block
+
+            tl.atomic_add(DW_ptr + col_offsets, dW_block, mask=mask)
+            tl.atomic_add(DB_ptr + col_offsets, dB_block, mask=mask)
```
Comment (Tcc0403) [inline src/liger_kernel/ops/backends/_ascend/ops/layer_norm.py:371]: Right, I didn't catch this part when reviewing rms_norm kernel. How's the performance with/without atomic_add?

File: src/liger_kernel/ops/backends/_ascend/ops/layer_norm.py [hunk 3]
```diff
@@ -0,0 +1,681 @@
...
+):
+    """NPU-optimized layer_norm backward kernel with column blocking."""
+    pid = tl.program_id(0)
+    num_progs = tl.num_programs(0)
+    num_col_blocks = tl.cdiv(n_cols, BLOCK_SIZE)
+
+    offsets = tl.arange(0, BLOCK_SIZE)
+    n_cols_inv = 1.0 / n_cols
+
+    for row_idx in tl.range(pid, n_rows, num_progs, num_stages=NUM_STAGES):
```
Comment (Tcc0403) [inline src/liger_kernel/ops/backends/_ascend/ops/layer_norm.py]: https://github.com/linkedin/Liger-Kernel/pull/1111
It seems num_stage doesn't affect the result on npu devices?

File: src/liger_kernel/ops/backends/_ascend/ops/layer_norm.py (hunk 1 above)
Comment (TianHao324) [reply src/liger_kernel/ops/backends/_ascend/ops/layer_norm.py]: When the NPU device is performing accumulation, if there are too many accumulation items, errors may occur.

File: src/liger_kernel/ops/backends/_ascend/ops/layer_norm.py (hunk 2 above)
Comment (TianHao324) [reply src/liger_kernel/ops/backends/_ascend/ops/layer_norm.py:371]: When using atomic_add, if multiple programs simultaneously update the same position of DW_ptr and DB_ptr, the atomic operation ensures the correctness of the accumulation and incurs a loss of some performance. However, if we do not use atomic_add (such as when n_cols is small (≤2048)), considering the UB limit of the NPU, we choose to trade off a smaller performance cost for memory.

File: src/liger_kernel/ops/backends/_ascend/ops/layer_norm.py (hunk 3 above)
Comment (TianHao324) [reply src/liger_kernel/ops/backends/_ascend/ops/layer_norm.py]: OK, I have completed the revision and I will pay attention to this issue in the subsequent PR.

**Comment by TianHao324** [comment]
@Tcc0403 Hello, are you a bit busy recently? May I ask if you have time to take a look at this PR as well as the others I submitted recently?

**Comment by Tcc0403** [comment]
> May I ask if you have time to take a look at this PR as well as the others I submitted recently?

Sure, I'll review your recent PRs. If you have any new PRs or get any updates on existing ones, feel free to ping me. I wasn't checking all PRs regularly, so I might miss your updates.

**Comment by TianHao324** [comment]
> Sure, I'll review your recent PRs. If you have any new PRs or get any updates on existing ones, feel free to ping me. I wasn't checking all PRs regularly, so I might miss your updates.

OK. Thank you for your explanation.

File: src/liger_kernel/ops/backends/_ascend/ops/layer_norm.py [hunk 4]
```diff
@@ -0,0 +1,673 @@
...
+        row_sum = 0.0
+
+        for col_block_idx in range(num_col_blocks):
+            col_start = col_block_idx * BLOCK_SIZE
+            col_offsets = col_start + offsets
+            mask = col_offsets < n_cols
+
+            X_block = tl.load(X_row_ptr + col_offsets, mask=mask, other=0.0, cache_modifier=".cg").to(tl.float32)
+
+            row_sum += tl.sum(tl.where(mask, X_block, 0.0))
```
Comment (Tcc0403) [inline src/liger_kernel/ops/backends/_ascend/ops/layer_norm.py]: We don't have to do `tl.where` since out of bound values are masked to 0.0 in the previous `tl.load`

File: src/liger_kernel/ops/backends/_ascend/ops/layer_norm.py [hunk 5]
```diff
@@ -0,0 +1,673 @@
...
+
+        for col_block_idx in range(num_col_blocks):
+            col_start = col_block_idx * BLOCK_SIZE
+            col_offsets = col_start + offsets
+            mask = col_offsets < n_cols
+
+            X_block = tl.load(X_row_ptr + col_offsets, mask=mask, other=0.0, cache_modifier=".cg").to(tl.float32)
+
+            X_centered = X_block - mean
+            var_sum += tl.sum(tl.where(mask, X_centered * X_centered, 0.0))
```
Comment (Tcc0403) [inline src/liger_kernel/ops/backends/_ascend/ops/layer_norm.py:163]: This `tl.where` is required since out of bound values in `X_centered` became `-mean`, which should be masked out again. Or it will result in incorrect `var` value.

File: src/liger_kernel/ops/backends/_ascend/ops/layer_norm.py [hunk 6]
```diff
@@ -0,0 +1,673 @@
...
+        grid_size = _compute_grid_size(n_rows, BLOCK_SIZE_M, num_cores)
+        DW_scratch = torch.empty((grid_size, n_cols), dtype=torch.float32, device=W.device)
+        DB_scratch = torch.empty((grid_size, n_cols), dtype=torch.float32, device=W.device)
+
+        n_cols_inv = 1.0 / float(n_cols)
+
+        _layer_norm_backward_kernel_no_tiling[(grid_size,)](
+            X,
+            X.stride(0),
+            X.dtype if hasattr(X.dtype, "name") else tl.float32,
```
Comment (Tcc0403) [inline src/liger_kernel/ops/backends/_ascend/ops/layer_norm.py]: In what scenario, would `X.dtype` have attribute `name`?

File: src/liger_kernel/ops/backends/_ascend/ops/layer_norm.py [hunk 7]
```diff
@@ -0,0 +1,673 @@
...
+        wdy_masked = tl.where(block_mask, wdy, 0.0)
+
+        c1 = tl.sum(x_hat_wdy_masked, axis=1) * n_cols_inv
+        c2 = tl.sum(wdy_masked, axis=1) * n_cols_inv
+
+        DX_f32 = (wdy - (x_hat * c1[:, None] + c2[:, None])) * rstd_rows[:, None]
+
+        # Store dX with coalesced memory access
+        DX_block_ptr = DX_ptr + row_idx[:, None] * DX_row_stride + col_offsets[None, :]
+        tl.store(DX_block_ptr, DX_f32.to(X_dtype), mask=block_mask)
```
Comment (Tcc0403) [inline src/liger_kernel/ops/backends/_ascend/ops/layer_norm.py]: I'm not sure about triton code would be compiled for npu device, but `tl.store` should implicitly typecast `DX_f32` to `DX.dtype` 


> value is implicitly broadcast to pointer.shape and typecast to pointer.dtype.element_ty. 

Ref: https://triton-lang.org/main/python-api/generated/triton.language.store.html

File: src/liger_kernel/ops/backends/_ascend/ops/layer_norm.py [hunk 8]
```diff
@@ -0,0 +1,673 @@
...
+# -----------------------------------------------------------------------------
+# Optimized Backward Kernel - No Tiling (for n_cols <= 2048)
+# -----------------------------------------------------------------------------
+
+
+@triton.jit
+def _layer_norm_backward_kernel_no_tiling(
+    X_ptr,
+    X_row_stride,
+    X_dtype: tl.constexpr,
```
Comment (Tcc0403) [inline src/liger_kernel/ops/backends/_ascend/ops/layer_norm.py]: You can also retrieve dtype by `X_ptr.dtype.element_ty` in triton code, so that we don't have to pass `X_dtype` in host side.

File: src/liger_kernel/ops/backends/_ascend/ops/layer_norm.py (hunk 4 above)
Comment (TianHao324) [reply src/liger_kernel/ops/backends/_ascend/ops/layer_norm.py]: I got it

File: src/liger_kernel/ops/backends/_ascend/ops/layer_norm.py (hunk 8 above)
Comment (TianHao324) [reply src/liger_kernel/ops/backends/_ascend/ops/layer_norm.py]: Thank you for your explanation. I didn't know this could be done before.

**Comment by TianHao324** [comment]
@Tcc0403 I have fixed the issues related to masks and type conversions. Are there any other questions?

File: src/liger_kernel/ops/backends/_ascend/ops/layer_norm.py (hunk 1 above)
Comment (Tcc0403) [reply src/liger_kernel/ops/backends/_ascend/ops/layer_norm.py]: I should've asked earlier. Could you provide detailed test cases it wasn't able to pass with using normal `torch.sum` only?

Comment (TianHao324) [reply src/liger_kernel/ops/backends/_ascend/ops/layer_norm.py]: Sorry, when I was trying to reproduce this issue just now, I found that there was no such error problem anymore. It might be caused by my previous incorrect operation. Currently, I have submitted a new modification to remove this part. Please have a look.
